post/2023/beyond-sf/index.qmd

```diff
@@ -0,0 +1,305 @@
+
+```{r}
+c(
+  st_sfc(pnt, pnt), 
+  st_sfc(pnt),
+  st_sfc(ln)
+)
+```
+
+Under the hood, this is just a list of `sfg` objects. Remember, it's not magic! 
+
+```{r}
+unclass(st_sfc(pnt, pnt))
+```
+
+Since `sfc` objects are vectors, they can be included as a column in a data frame. 
+
+```{r}
+df <- data.frame(
+  geo = st_sfc(pnt, pnt)
+)
+
+df
+
+class(df)
+```
+
+## Geometry and data.frames
+
+Having geometry included in a data frame was a huge win for the R community because this means that attributes can be included along with geometries. Further, geometries in a data frame means that they are dplyr compatible. 
+
+For example we can create a vector of points from the `x` and `y` columns from the `diamonds` dataset in `ggplot2`.
+
+```{r}
+data(diamonds, package = "ggplot2")
+
+pnts <- purrr::map2( # <1>
+  diamonds$x, 
+  diamonds$y, 
+  function(.x, .y) st_point(c(.x, .y)) # <2>
+) |> 
+  st_sfc() # <3>
+```
+1. We iterate over both the x and y columns in the diamonds dataset returning a list
+2. Each value of `x` and `y` are accessed through the placeholder `.x` and `.y` based on their position in the `map2()` function. `st_point()` takes a length 2 numeric vector and returns a single `sfg` `POINT` object.
+3. We pass the list of `sfg` objects to create an `sfc`
+
+We can included this in a data frame 
+
+```{r}
+library(dplyr, warn.conflicts = FALSE)
+
+dmnd <- diamonds |> 
+  select(price, clarity) |> 
+  bind_cols(geometry = pnts)
+
+head(dmnd)
+```
+
+Now we have a tibble with price, clarity, and geometry! Huge win! What if we wanted to calculate the average price by clarity and keep the geometries?
+
+```{r}
+dmnd |> 
+  group_by(clarity) |> 
+  summarise(avg_price = mean(price))
+```
+
+Unfortunately, we lose the geometry just like we would for any other column that wasn't included in the `summarise()` call. To keep it, we would need to perform an operation on the geometry itself. 
+
+```{r}
+dmnd |> 
+  group_by(clarity) |> 
+  summarise(
+    avg_price = mean(price),
+    geometry = st_union(geometry)
+  )
+```
+
+Wouldn't it be nice if the geometry knew to do that? Well, that is exactly what `sf` objects are.
+
+## `sf` objects
+
+`sf` objects are just data frames with a geometry column that is *sticky* and *smart.* We can create an `sf` object if an `sfc` column is present in a data frame by using `st_as_sf()`. 
+
+```{r}
+dmnd_sf <- st_as_sf(dmnd)
+```
+
+Doing this creates an object of class `sf`. The two things that make an `sf` object so special are the class `sf` and the attribute `sf_column`. 
+
+```{r}
+attr(dmnd_sf, "sf_column") # <1>
+```
+1. `attr()` lets us access a single attribute by name
+This attribute tells us which column is the geometry. Because we have this `sf` can implement its own methods for common functions like `select()`, `mutate()`, `aggregate()`, `group_by()`, etc which always keep the `attr(x, "sf_column")` attached to the data frame. 
```

Comment:
```suggestion
This attribute tells us which column is the geometry. Because we have this, `sf` can implement its own methods for common functions like `select()`, `mutate()`, `aggregate()`, `group_by()`, etc., which always keep the `attr(x, "sf_column")` attached to the data frame. 
```